### PR TITLE
Flex menus

### PR DESCRIPTION
**Pull request by joekur**
Hey, I've been working on getting this menu system set up where you can add more than just pages to a menu. I've got it to the point where you can configure it to accept any refinery file type (pages, blog posts, resources, etc) as well as make links to custom URLs. Here are some screenshots to get the idea across:

Edit Page:
Edit Page
If you're familiar with wordpress, you'll see where I got the inspiration :)

Tree Index Page:
Index Page

I think the biggest question is whether you want to extend your gem to include this functionality, or you think this is better suited as its own extension. I'm happy either way!

### CODE REVIEW

**Comment by jokklan** [comment]
This looks awesome! I will look through soon, when i get the time

File: app/models/refinery/page_position.rb [hunk 1]
```diff
@@ -10,15 +11,70 @@ class PagePosition < Refinery::Core::BaseModel
...
+      refinery_resource_id.present? && refinery_resource_type.present?
+    end
+
+    def resource
+      return page if refinery_page_id # for now, until we phase out 'refinery_page_id'
+      return nil if custom_link?
+      resource_klass.find(refinery_resource_id)
+    end
+
+    def resource_url
```
Comment (joekur) [inline app/models/refinery/page_position.rb:60]: This is still incomplete - we need a way to incorporate the different routing of the objects, possibly determined by the configurations.

**Comment by jokklan** [comment]
Im sorry that i haven't answered, in i still have no time to look this through before next week. However it's looks very good, and i look forward to pull this into the project !  

I know that the project lacks tests already, but will you have time to include tests in your pull request?

File: app/models/refinery/page_position.rb (hunk 1 above)
Comment (jokklan) [inline app/models/refinery/page_position.rb:60]: Im pretty sure that refinery have some methods or helpers for this..

**Comment by joekur** [comment]
Yes, I definitely plan on including tests; no guarantee it'll be this week though.

I did want to discuss with you about whether we want to have two separate pages (shown by the two screenshots above), or whether it would be simplest to include the nested sortable behavior in the edit page and remove the index page altogether.

**Comment by jokklan** [comment]
I have to clone it and look at it myself before im sure, but i think it will be best to include the nested sortable behavior in the edit page, and remove the index page, or maybe just have a index over pages without any menu stuff, much like they have in wordpress.

**Comment by jokklan** [comment]
@joekur do you think it would be best to have menus in it's own tap (again like wordpress), or should it be a submenu under Pages like now ?

**Comment by jokklan** [comment]
The more i work with this, the more i lean against a "wordpress-like" solution. 

I think it would be best with a seperate tap for "Page Menus" (maybe change to just Menus) with a list of menus, each editable with your flex_select form. Then i think we should add nesting of menu items to this form, and don't have a seperate interface for this.

The Pages index page should not be edited at all, so their will still be the old pages layout with the tree index page, in it's own tap.

This solution have some advantages:
1. It doesn't change as much of the original pages engine, and will thereby be much easier to upgrade in the future + i think it's an advantage in itself, not to change how RefineryCMS already works.
2. We still have the normal parent-children tree index page and this is actually also how they do in wordpress. This gives you options like breadcrumbs, nested page menus and so forth.
3. We have anything menu related in one tap, pretty much like a normal extension, with a list of menus, that can be edited separately. Maybe later we can update this, so the menu index pages will be more informative then just a list, as their properly won't be more then a few menus in a single application.

What do you think @joekur? Any comments, thoughts and ideas is appreciated :) !

**Comment by joekur** [comment]
I think I agree with everything you just said :)

Potentially the menus index page could just be the edit page for the first menu, with links to switch to the other menus (and maybe to create a new menu if we want this to be available on the front end). I think this again is similar to how wp does it? I'm not sure what other info we could show about the menus.

Btw, sorry I've been so slow with writing the tests. I got frustrated getting the js driver to play nicely with the menu edit page. I want to try changing the javascript driver from selenium to capybara and see if that fixes the problem (and it's also headless and faster).

**Comment by jokklan** [comment]
Sure no problem, i to have been busy too. 
And yes i got the exactly same problem when i tried to run the tests... Don't know what causes it yet.
But i think i will try to move the views over in it's own tap, and add tree-sorting functionality to the form page of page menus, and you can then look it over and see what you think :)

**Comment by jokklan** [comment]
I have partly integreted the tree-nesting functionality now. However to show the nesting in the view (the form), do I either need to update your js rendering functions, or change to normal partial rendering. So is it a problem if I change it to normal rendering, or was their a special reason for doing this in js? Or do you maybe have a better solution?

**Comment by joekur** [comment]
I decided to render on the front end since I was allowing you to create new links on the front end. Do you have preference against this approach? Regardless, the front end will need to know how to render a new link, so for the sake of simplicity, it seemed best to just give data to to the client and let it handle initially rendering everything. I do wish with this method though that we had a templating engine (like handlebars, haml-coffee-assets), because yes those render functions look a little ugly :P

**Comment by jokklan** [comment]
Personally i don't like to have templates outside the view folders, or render a whole html template in the js language. But a js/html template engine would of course solve this. 
Other solutions could be ajax calls, or a hidden empty block that could be copied and filled out by js... Im not sure however what would be the best approach, but i think it would be prettier if we somehow could remove the view code out of javascript.

**Comment by jokklan** [comment]
I have added a new pull request (#9) , with some changes. I would love to hear your thoughts about the changes.

**Comment by jokklan** [comment]
I have added it as a pull request to your own repository, so you can see my changes alone :) (joekur/refinerycms-page-menus#1)

**Comment by jokklan** [comment]
Actually, after working with this for some time, i agree with your method of rendering in javascript only. However it would be better if we could use a template engine as you suggested :)! 
Do you think that you have time to implement that?

**Comment by jokklan** [comment]
Sorry for changing opinion all the time... But i have encountered a new problem:
I have added the ability to nest page_positions in the menu (maybe we should change the name to links for the model) in the flex_select form. However to store the positions of the links, does the links have to be updated ie. they have to be saved to the db first. So when adding new links to a menu, you have to save the menu before you can change the structure. A solution to this would be using ajax calls when adding links to the menu, and that would of course change how the menu is rendered...

**Comment by jokklan** [comment]
Another solution, would be to change the current nesting implementation (it's is currently working like pages with a reorder button), so reordering always is enabled, and the structure is saved when saving the menu.

**Comment by joekur** [comment]
Yeah, I haven't had time to look at your code yet, but I'm thinking we'll probably have to roll our own reorder/nesting functionality since we will allow creating new links and reordering them on the same page.

I can look back into the templating thing, but I'm honestly thinking it may not be worth it since we have just two templates we're rendering and we'd be adding a dependency. But I totally understand your frustration with the way I was doing it with strings :)

**Comment by jokklan** [comment]
As you can see have i pulled a ajax solution, that is more simple in several ways. However the user experience is of course a bit slower, but i don't think it's a problem. The best solution of course would be our own reorder/nesting functionality as you suggested @joekur, but i think it's to much work. 
However if you find that it's worth it, will i be happy to pull it in, as long as it's not MUCH more complex that the current ajax solution :)

However thanks for your help, it have really changed this engine to something actually useful :+1:  !

**Pull request merged**